hal/src/i2c.rs

```diff
@@ -0,0 +1,448 @@
+//! # Inter-Integrated Circuit (I2C) bus
+
+use core::convert::TryFrom;
+
+use crate::{
+    embedded_hal::blocking::i2c::{Read, Write, WriteRead},
+    gpio::{OutputType, Pull},
+    pac::{rcc::ccipr::I2C3SEL_A, I2C1, I2C2, I2C3, RCC},
+    rcc::{pclk1_hz, sysclk_hz},
+};
+
+use embedded_time::{fixed_point::FixedPoint, rate::*};
+
+/// I2C error
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[non_exhaustive]
+pub enum Error {
+    /// Arbitration loss
+    Arbitration,
+    /// Bus error
+    Bus,
+    /// Bus busy
+    Busy,
+    /// Not Acknowledge received
+    Nack,
+    /* Overrun, // slave mode only
+     * Pec, // SMBUS mode only
+     * Timeout, // SMBUS mode only
+     * Alert, // SMBUS mode only */
+}
+
+/// I2C peripheral operating in master mode
+pub struct I2c1<PINS> {
+    i2c: I2C1,
+    pins: PINS,
+}
+
+pub struct I2c2<PINS> {
+    i2c: I2C2,
+    pins: PINS,
+}
+
+pub struct I2c3<PINS> {
+    i2c: I2C3,
+    pins: PINS,
+}
+
+macro_rules! busy_wait {
+    ($i2c:expr, $flag:ident, $variant:ident) => {
+        loop {
+            let isr = $i2c.isr.read();
+            let icr = &$i2c.icr;
+
+            if isr.arlo().is_lost() {
+                icr.write(|w| w.arlocf().clear());
+                return Err(Error::Arbitration);
+            } else if isr.berr().is_error() {
+                icr.write(|w| w.berrcf().clear());
+                return Err(Error::Bus);
```

Comment:
The bus error should be ignored in master mode, from the [erratum](https://www.st.com/resource/en/errata_sheet/es0500-stm32wl55xx-stm32wl54xx-device-errata-stmicroelectronics.pdf):

> In master mode, a bus error can be detected spuriously, with the consequence of setting the BERR flag of the I2C_SR register and generating bus error interrupt if such interrupt is enabled. Detection of bus error has no effect on the I2C-bus transfer in master mode and any such transfer continues normally.

> If a bus error interrupt is generated in master mode, the BERR flag must be cleared by software. No other action is required and the ongoing transfer can be handled normally.